Restructuring IMAU-ICE

carolinevancalcar opened the pull request:
IMAU-ICE is fully restructured to solve the following two problems:
1. Not all modules were updated before the new ice sheet geometry was computed. The new version runs all the modules during initialization in the same order as the main model loop so that the model is fully consistent.
2. Differences between a restart run and a continuous run were very big. Now, the model can be restarted and provides the exact same result as the continuous run. You can use any combination of time stepping and ice dynamic method that is available. It also possible to use a variable time step.

**Important note**: timesteps should always be chosen carefully!
- All timesteps should always be a binary number from the power of 2, for example 0.125, 0.25 or any integer.
- When using the SMB model IMAU-ITM, dt_SMB_config should be set to 1 year.

This merge also contains many small fixes, of which we list the most important ones below.
- The model can now read bedrock deformation (computed by an external GIA model) from a file and applies the deformation linearly at each time step of the ice model. “choice_GIA_model_config” can be set to “externalGIA” to use this option and “filename_dHb_externalGIA_config” is used to set the name of the file. The new config option “do_write_relative_ice_load” computes relative ice load compared to the GIA reference ice load so that it can be written to the help field file (set to false by default). The relative ice load is used to compute bedrock deformation by an external GIA model or to see the ice load that ELRA uses to compute bedrock deformation.
- Add new config variable “dt_output_regional_scalar_config” (previously do_write_regional_scalar_every_timestep). This variable is set to 5000 by default. This config determines at which time interval the regional scalar output should be written.
- Several CALL sync are added in the code to fix stochastic behavior of the model. The model is now fully deterministic when PETSc is used.
- The lapse rate correction in the isotope module to obtain temperature at sea level was incorrectly applied, making it colder instead of warmer. This is now fixed.

Caution!
- Two equations to solve velocities with the DIVA method were wrong and are corrected. However, these equations were of little influence on the resulting velocities. The velocities are well computed in idealized experiments but are too slow in realistic experiments for Antarctica. I strongly advice not to use DIVA for realistic simulations of Antarctica.
- Note that the SELEN module is not tested after restructuring the model. This should be done before using SELEN.

carolinevancalcar committed:
Remove the half of the wind

Meike tested to half the wind, this is now removed from the code.

carolinevancalcar committed:
[WIP] Shuffle the order of the main loop

The order is now correct but the timestepping is not correct yet.
Added CvC in comments to recognize changes in timestepping.

carolinevancalcar committed:
[WIP] Main model loop contains the correct order and output is written well

Ice thickness is not correct, something is broken.

carolinevancalcar committed:
[WIP] The great shuffle is working

Seperate the computation of t_next from the action of setting the flags
to run the models. First compute t_next, then set time to time+t_next
and then set the flags to run each model to true or false. The shuffle
is working well now. However, the restart of the velocities is not correct
and the velocities written to the help_field and restart are very different
at the same step in time.

carolinevancalcar committed:
[WIP] Call velocities each time step, 5 viscosity iterations, Ti restart fix

The solve_SSA and solve_SIA subroutine are called every time step.
The viscosity iteration is done 5 times each time its called.
The temperature model is only run in the initialisation when its
not a restart run.

carolinevancalcar committed:
[WIP] Fix position of next timestep determinination

The time step of the next step should be determined before computing
the ice thickness for the next time step. That is now fixed. Concerning
the restart, all files are correctly read in for the restart run. After
initialising, the difference between the "no restart" run and the restart
run is 1.5 millimeter in ice thickness at 4 gridcells in the peninsula
and maximum 80 m/year in velocity at an ocean gridcell.

carolinevancalcar committed:
[WIP] Clean version

restart works if you use correct timesteps. 0.25 or multiples of that is fine.

carolinevancalcar committed:
Add print statements to checksum

print main variables to terminal to find where call sync is missing.

carolinevancalcar committed:
[WIP] Fix parallelisation problem

There was a missing call sync in the determination of the grounded
fractions in general_ice_model_data_module that caused stochastic
behavior in the model. Also added some other call syncs that didn't
make a big difference but were missing anyway.

FrankaJes committed:
[WIP] Velocity restart update SSA and DIVA

Implemented the IF statement to skip the viscosity iteration in solve_DIVA when reading in velocities from the restart. Need to figure out if this is possible because it might need the viscosity field as well which is computed during the iteration.

Implemented some IF statements in order to be able to read and write velocity data from and to output when running SSA only. (Was only implemented for SIA/SSA and DIVA)

FrankaJes committed:
Update tabs in ice_velocity_module

Old tabs raised an error during compilation

carolinevancalcar committed:
Fix stochastic behavior of SOR

Tijn changed some lines of code in the SOR solver to make it
fully deterministic.

carolinevancalcar committed:
Add missing call sync to SOR subroutine

carolinevancalcar committed:
Fix race condition, Hb shuffle and Hi initialisation

Found race condition in derivatives_and_grids_module. Before mapping
from one grid to the other, all processors should be in sync.

Hb was computed and updated after the ice thickness and the mask but
should be done before. Also, dHb_dt should be computed at the end of
the initialisation.

During initialisation, Hs, Hi and Hb were recomputed based on Hb + dHb
in the adapt_initial_geometry_from_restart_file subroutine. When dHb is
zero, this caused and error in ice thickness of 10^-13. When dHb is nonzero,
this is incorrect following the new shuffle method. This whole subroutine
can be skipped and is commented now.

carolinevancalcar committed:
Fix restart for IMAU-ITM model

IMAU-ITM was reading firn depth and melt from the previous year
from the restart file. However, the restart file contained the newest
firn depth and melt, not the one from the previous year. To restart,
you actually need the one from the previous year. So I added a variable
to contains the firn depth and melt of the previous year for the restart
file. Note that dt_SMB_config should be set to 1 year to make this model
work well in the way it is written now. There is room for improvement on
this part.

carolinevancalcar committed:
Use same structure for ELRA and external GIA model

The bedrock elevation is now updated in the same way between ELRA
and an external GIA model.

carolinevancalcar committed:
Restart ELRA with dHb_dt instead of dHb

ELRA is now restarting with dHb_dt instead of dHb because the dHb
is dependent on the size of previous timestep, which is unknown in
a restart run. Now it is restarting from dHb_dt and then dHb is computed
from that.
Still need to clean up the code and create a do_restart if statement
when reading dHb_dt. Now it reads restarting with dHb_dt when ELRA is
turned on and when restart with velocities is set to true. But you always
want to restart with velocities so for the work in progress its okay.

carolinevancalcar committed:
Note: dt_bedrock and dt_SMB (with IMAU-ITM) should be 1 year

Because the restart model doesn't know when the next time should be
that the models should be updated, the restart works only if you
simulate bedrock and imau-itm each year.

carolinevancalcar committed:
Working version! Big cleanup

This versions works for projections, Franka's idealised setup and
the spinup. All comments are cleaned up and removed CALL syncs
Gaussian floodfill subroutine since that subroutine should only
be called by de master.

Users of IMAU-ICE should use particular timesteps for depending on
the application. For a perfect restart, set dt_min and dt_max to the
smallest other timestep that you choose for the modules. When using
the SMB model IMAU-ITM, use a time step of 1 year. All timesteps
should always be a binair number, for example 0.125, 0.25 or
any other integer.

carolinevancalcar committed:
Fix dt=0 in last time step

t_next was limited by t_end, causing dt to be zero for the last
time step. Therefore, Hi would be divided by zero to compute dHi_dt,
which resulted in NaN values. It wasn't needed to limit t_next so
that part is removed.

carolinevancalcar committed:
[WIP] Add print statements to check PC method

PC method computes a critical time step with many decimals. Therefore,
the time of the main model loop never arrives at the t_next_SIA but goes past
that time step. after three timesteps, the velocities are not computed anymore
and region%dt stays at the minimum dt. Also, the output is not written because
the main time loop skips the t_next_output.

carolinevancalcar committed:
Update region%dt in the pc method

Call the subroutine determine_timesteps in the pc method to update
region%dt. This was done in the direct method but was missing in the
pc method.

carolinevancalcar committed:
Create config to set dt for regional scalar output and resource tracking

The config variable "do_write_regional_scalar_every_timestep_config" is
changed to "dt_output_regional_scalar_config" so that you can choose the
timestep for which to write regional scalar output.

The new config variable "do_write_resource_tracking_output_config" sets
whether to write the resource tracking file or not.

This commit also contains some changes in documentation that was needed
after the shuffle.

FrankaJes committed:
Fix DIVA in ice_velocity_module

Solved inconsistencies between IMAU-ICE and UFEMISM DIVA solver. Main differences were in the subroutines calc_basal_velocities and calc_3D_horizontal_velocities_DIVA. In calc_basal_velocities, I implemented the equation 32 from Lipscomb (instead of Goldberg 2011).
In the calc_3D_horizontal_velocities_DIVA, I added an IF statement for no_sliding case and made sure the equations are consistent with Lipscomb eq. 29.
Also corrected and added references to Lipscomp et al. (2019).

FrankaJes committed:
Update netcdf_output_module

Fixed DIVA velocity output

carolinevancalcar committed:
[WIP] predictor/corrector works with SIA/SSA

dHb_dt and dHi_dt are now initialised in the new subroutine called
"initialise_changing_geometry_from_restart_file" so that dHs_dt is
computed from the correct geometry (before dHi_dt was zero at the
start of the restart run). Attention: I should implement that dHi_dt
is the second last timestep when pc is used and the last timestep
when direct is used.

When the  pc method is used, the second last timestep u_vav_cx and
v_vav_cy are safed to the restart file. When direcht method is used,
these variables are safed for the last time step only when DIVA is
used.

This commit contains many print statements because I still need to
test pc in combination with DIVA and I need to conduct a final test
for all combinations.

carolinevancalcar committed:
[WIP] Fix dHi_dt for direct

Now dHi_dt is also initialised in direct method but is not necessary,
just for consistency.
Also added missing END IF statement.

carolinevancalcar committed:
Merge branch 'project/shuffle' of https://github.com/IMAU-paleo/IMAU-ICE into project/shuffle

carolinevancalcar committed:
[WIP] Working DIVA for continues run

The restart with DIVA does not work yet. I am working on that.
This commit also contains a fix in the ice_dynamics module for
writing regional scalar output. The t_next and
do_write_output_regional_scalar were not updated at each loop.

This versions still contains many commented print statements because
I am testing the restart.

carolinevancalcar committed:
Working restart for all dynamics and timestepping variations

The restart works for direct and pc in combination with SIA, SSA,
SIA/SSA and DIVA. This commit also contains a big cleanup of print
statements.

The restart still only works when dt_min and dt_max are equal and set
to the smallest dt of the other modules. For example if BMB is 0.25 and
the other dt's are bigger, choose 0.25 for dt_min and dt_max.

Commit:
Reading 0D fields from restart

New subroutine added (read_field_from_file_0D),  which should be used to either (1) load a single value or (2) load one value at a certain point in time (e.g., from a restart).

Also fixed writing 0D fields from the predictor corrector to the restart.

NOTE: There are some comments (look for WIP or MS) for variables that cannot yet be loaded due to data-type issues!

Commit:
Improve comments on writing 1D and 0D data

Improved comments to prevent further confusion between writing time_history_1D and 0D data. And removing some print statements.

Commit:
SMB dummy fix for climate matrix

The SMB_module needs a MeltPreviousYearforrestart and wFirnDepthforrestart field. These fields were not yet part of the SMB_dummy in the climate matrix method, causing a crash in the SMB_module. This has now been fixed.

Commit:
Restarting with Isotope History Fix

Restarting from an isotope simulation gave segmentation faults.

This has now been fixed. And, instead of a segmentation fault it should now tell you that a variable cannot be found.

(This bug was already there before the shuffle)

Commit:
Isotopes correctly calculated during first time-step

Isotopes at first time-step were too low (minimum in all locations).

By calling the isotopes after the run_ice_model during initialisation, this problem is fixed.

Commit:
Fixing the isotope module lapse rate

The lapse rate correction in the isotope module to obtain temperature at sea level was incorrectly applied, making it colder instead of warmer.

Isotope module is now called after the running ice dynamics.

Some clean-up of comments in the isotope module.

This means the isotope model should be fully working again.

carolinevancalcar committed:
Clean up

Clean up of some unused variables

carolinevancalcar committed:
Merge branch 'project/shuffle' of https://github.com/IMAU-paleo/IMAU-ICE into project/shuffle

carolinevancalcar committed:
Small fixes to read and write timevariables from/to the restart file

carolinevancalcar committed:
Restart with variable time step

The model can now restart with a variable time step using all
available combinations of ice dynamics and time stepping methods.

This commit also contains some cleaning up of variables in the
restart file.

carolinevancalcar committed:
Merge pull request #53 from IMAU-paleo/project/shuffle_isotopes

Project/shuffle isotopes

carolinevancalcar commented:
@TijnBerends In cursief mijn antwoord.

1.	bedrock_module, line 96: waarom gebruik je Hi( t+dt) om de surface load uit te rekenen i.p.v. Hi( t)? _De bedrock update komt voor het updaten van de ijsdikte maar je wil al wel de deformatie voor de komende tijstap uitrekenen dus vandaar dat je de load uitrekent voor Hi_tplusdt._
2.	ice_velocity_module, line 183 - 239: zijn hier enkel twee spaties toegevoegd, wellicht als overblijfsel van een IF statement dat je geprobeerd hebt? _Niet gezien dat er spaties zijn overgebleven, hier heeft inderdaad een tijd een if statement omheen gestaan maar was uiteindelijk niet nodig._
3.	ice_velocity_module, line 312 - 372: zelfde, maar hier staat er wel een IF statement omheen, is dat de bedoeling? _Goed gespot. Dat if statement doet niks meer dus beter om weg te halen._
4.	utilities_module, line 515 - 726: dit zijn de aanpassingen aan de SOR die ik toen gauw-gauw heb geschreven om te checken of her een parallelisatieprobleem in zat. Dit kan op zich weg - weliswaar krijg je dan dus wel weer dat runs met meer dan één core niet deterministisch zijn, maar de SOR zou wel een eind sneller moeten worden. _Is goed. Als PETSc sneller, stabieler en deterministisch is dan is het altijd beter om PETSc te gebruiken toch?

Is het handig als ik punt 3 aanpas of als jij dat doet als reviewer? Ik snap nog niet helemaal hoe pull request werkt volgensmij._Kan jij punt 4 aanpassen?

carolinevancalcar committed:
Remove unnecessary if statement in velocity module

There was an old if statement in the velocity module that prevents
the viscosity iteration to run when the model in initialising from
a restart but in the final method used to restart, that subroutine
is not called for at all when the model is initialising from a
restart.

carolinevancalcar committed:
Update config files for spinup and projection

Adjusted variables to standard values from their respective papers
and updated spinup variables so that it produces a spinup for
Antarctica that provides an equilibrium state for about 1000 years.
For longer time scales, this config file should be adjusted.

carolinevancalcar commented:
UPDATE:
I have added two template configuration files, one for an Antarctic spinup and one for an Antarctic projection.

TijnBerends committed:
Removed SOR parallelisation check

Removed the temporary change to the SOR matrix solver where it was effectively reduced to a Gauss-Seidel solver (by only using old values of x, instead of the most recent) to check for parallelisation issues that were troubling us back then.

TijnBerends reviewed (approved):
Lijkt mij goed zo.

MScherrenberg commented:
Ziet er goed uit. Het is gelukt om te compileren en te runnen. Daarnaast heb ik nog een test gedaan met 2 ijskappen (Eurazië en Noord Amerika) en dat lijkt het te doen.

Een paar kleine opmerkingen:

- Je zegt dat IMAU-ITM elk jaar moeten runnen omdat het anders problemen zorgt voor de restart (of ook voor een continue simulatie?). Heb je hiervoor ook een waarschuwing / hardcode / crash ingebouwd als een gebruiker een ander nummer dan 1 invult? Het is namelijk iets wat makkelijk wordt vergeten.

- Hetzelfde gaat ook op voor tijdstappen, want deze moeten altijd een binary nummer zijn. Dit zou ook in de configuration module moeten worden beschreven. Want dan is het heel makkelijk te vinden wat de tijdstap variabele betekenen, en dus ook waar je op moet letten.

- Op lijn 110 van IMAU-ICE_main staat: 
CALL apply_SELEN_bed_geoid_deformation_rates( region) ! Be aware although SELEN is chosen to be the GIA model, it is never run.

Is het niet beter als we daar gewoon een crash van maken? Want het zit in een if statement die checkt of je SELEN gebruikt.

carolinevancalcar committed:
Built in crash calls to prevent mistakes

Built in crash calls if timestep for IMAU-ITM is not 1.0 and if
SELEN is used. Also add explanation about timesteps to configuration
file and add to the main model to always write outputfiles at the
end of the simulation.

carolinevancalcar commented:
> Ziet er goed uit. Het is gelukt om te compileren en te runnen. Daarnaast heb ik nog een test gedaan met 2 ijskappen (Eurazië en Noord Amerika) en dat lijkt het te doen.
> 
> Een paar kleine opmerkingen:
> 
> * Je zegt dat IMAU-ITM elk jaar moeten runnen omdat het anders problemen zorgt voor de restart (of ook voor een continue simulatie?). Heb je hiervoor ook een waarschuwing / hardcode / crash ingebouwd als een gebruiker een ander nummer dan 1 invult? Het is namelijk iets wat makkelijk wordt vergeten.
> * Hetzelfde gaat ook op voor tijdstappen, want deze moeten altijd een binary nummer zijn. Dit zou ook in de configuration module moeten worden beschreven. Want dan is het heel makkelijk te vinden wat de tijdstap variabele betekenen, en dus ook waar je op moet letten.
> * Op lijn 110 van IMAU-ICE_main staat:
>   CALL apply_SELEN_bed_geoid_deformation_rates( region) ! Be aware although SELEN is chosen to be the GIA model, it is never run.
> 
> Is het niet beter als we daar gewoon een crash van maken? Want het zit in een if statement die checkt of je SELEN gebruikt.

Ik heb al je

carolinevancalcar committed:
Merge branch 'project/shuffle' of https://github.com/IMAU-paleo/IMAU-ICE into project/shuffle

carolinevancalcar commented:
> Ziet er goed uit. Het is gelukt om te compileren en te runnen. Daarnaast heb ik nog een test gedaan met 2 ijskappen (Eurazië en Noord Amerika) en dat lijkt het te doen.
> 
> Een paar kleine opmerkingen:
> 
> * Je zegt dat IMAU-ITM elk jaar moeten runnen omdat het anders problemen zorgt voor de restart (of ook voor een continue simulatie?). Heb je hiervoor ook een waarschuwing / hardcode / crash ingebouwd als een gebruiker een ander nummer dan 1 invult? Het is namelijk iets wat makkelijk wordt vergeten.
> * Hetzelfde gaat ook op voor tijdstappen, want deze moeten altijd een binary nummer zijn. Dit zou ook in de configuration module moeten worden beschreven. Want dan is het heel makkelijk te vinden wat de tijdstap variabele betekenen, en dus ook waar je op moet letten.
> * Op lijn 110 van IMAU-ICE_main staat:
>   CALL apply_SELEN_bed_geoid_deformation_rates( region) ! Be aware although SELEN is chosen to be the GIA model, it is never run.
> 
> Is het niet beter als we daar gewoon een crash van maken? Want het zit in een if statement die checkt of je SELEN gebruikt.

Goede suggesties! Ik heb het in de code gezet.

ps. ik zag ook dat het eind van de run_model subroutine (nadat de loop afgelopen is) nog een beetje rommelig eruit zag dus ik heb dat nog opgeruimd en gecodeerd dat output altijd weggeschreven wordt aan het eind van een run.

MScherrenberg reviewed (approved):
Prima, ik keur het goed. Dank je voor de aanpassingen!
